src/jupyterlabcontroller/models/domain/rspimage.py

```diff
@@ -0,0 +1,331 @@
+            Resulting image object.
+        """
+        return cls(
+            registry=registry,
+            repository=repository,
+            digest=digest,
+            **asdict(tag),
+        )
+
+    @property
+    def is_possible_alias(self) -> bool:
+        """Whether this tag could be an alias."""
+        return self.image_type in _ALIAS_TYPES
+
+    @property
+    def reference(self) -> str:
+        """Docker reference for this image."""
+        return f"{self.registry}/{self.repository}:{self.tag}"
+
+    @property
+    def reference_with_digest(self) -> str:
+        """Docker reference for this image, with the digest."""
+        return f"{self.registry}/{self.repository}:{self.tag}@{self.digest}"
+
+    def resolve_alias(self, target: RSPImage) -> None:
+        """Resolve an alias tag with information about its target.
+
+        If we discover the target tag of an alias tag, we can improve the
+        alias tag's display name and cycle information using the information
+        of the underlying tag. This normally happens when ingesting a set of
+        images, including alias images, into an `RSPImageCollection`.
+
+        If the tag was previously an unknown tag but we found another tag with
+        the same digest, assume it is an alias tag and upgrade it.
+
+        Parameters
+        ----------
+        target
+            Another tag with the same digest.
+
+        Raises
+        ------
+        ValueError
+            If this image has a type other than unknown or alias. (Ideally
+            this should be represented in the type system, but that code is
+            tedious and doesn't add much value.)
+        """
+        if not self.is_possible_alias:
+            raise ValueError("Can only resolve alias and unknown images")
+        self.image_type = RSPImageType.ALIAS
+        self.alias_target = target.tag
+        target.aliases.add(self.tag)
+        base_display_name = self.tag.replace("_", " ").title()
+        self.display_name = f"{base_display_name} ({target.display_name})"
+        self.cycle = target.cycle
+
+
+class RSPImageCollection:
+    """Provides operations on a collection of `RSPImage` objects.
+
+    Parameters
+    ----------
+    images
+        `RSPImage` objects to store.
+    """
+
+    def __init__(self, images: Iterable[RSPImage]) -> None:
+        self._by_digest: dict[str, RSPImage]
+        self._by_tag_name: dict[str, RSPImage]
+        self._by_type: defaultdict[RSPImageType, list[RSPImage]]
+
+        # Unresolved aliases by digest that are not in _by_digest.
+        self._unresolved_aliases: defaultdict[str, list[RSPImage]]
+
+        # Ingest all images.
+        self._replace_contents(images)
+
+    def add(self, image: RSPImage) -> None:
+        """Add an image to the collection.
+
+        Parameters
+        ----------
+        image
+            The image to add.
+        """
+        # If we're adding a non-alias image and we have unresolved aliases for
```

Comment:
Oh, that's nice.  It should be rare, so the expensive reindex should not happen with much frequency.